### ISSUE TITLE
Integrate AI Art through AI Horde

### ISSUE THREAD

**Issue by db0** [wesnoth/wesnoth#8264]
### Describe the desired feature

The [AI Horde](https://aihorde.net) is a FOSS, Crowdsourced service which can allow anyone to request Generative AI Art and text. There is opportunity for FOSS games like this to improve their look and provide unparalleled customization to the player by allowing integration with it.

Some possibilities to explore:

* Allow players to generate different unit portrait art per game.
* Allow players to stylize the unit sprites per game.
* Allow the players to customize their prompt styles which are used to generate art for their units. So, say one could generate armies which have a "gothic" look, while another could have armies with a more "steampunk" appearance. By utilizing things like img2img and controlnets you could generate unit portraits which closely follow the existing art frame you had until now, but would change in the style to match the player's suggestion, keeping things unique, but still recognisable.
* Alternatively provide a few "official" styles to choose from, so you retain some control of the art direction, but still create fresh looks each time.

The AI Horde also provides LLM functionality, which could also be utilized, for example to change the flavor text of the various common encounters to keep things a bit fresh.

All this could be optional from each player. So some players could opt to keep using the vanilla art and even avoid seeing custom art from their opponents.

The AI Horde provides a completely free and fully documented [REST API](https://aihorde.net/api) one can integrate with which can be used in realtime for free. 

Disclaimer: I'm the lead developer of the AI Horde and I'd love to see it used to improve the looks of games which historically have been struggling for art assets. I originally started it to be able to use Generative AI [for my own FOSS game](https://github.com/db0/hypnagonia/). Feel free to contact me for any assistance you'd require for this. I'm available in github, but also discord and matrix if you prefer.

**Comment by Pentarctagon** [comment]
At a minimum, this is a non-starter until the legal status of these image/text generators is clearly defined.

**Comment by db0** [comment]
Currently this is completely legal. Open lawsuits don't make something illegal.

**Comment by knyghtmare** [comment]
I am mentioning the three members to handle this issue whatever the outcome:
@Pentarctagon - The PM
@octalot - The Code person
@doofus-01 - the art director (units/portraits)

**Comment by Pentarctagon** [comment]
If their legality were clearly defined, the lawsuits would've been thrown out already.

**Comment by db0** [comment]
That's not true at all. Lawsuits are opened and lost all the time. The legality of Linux itself was challenged over many years by multiple Unix companies. It would be absurd to claim people should stop using Linux until those were resolved.

Anyway, as I said, this is about adding just novel code to integrate with an api. There's no potential issue here.

**Comment by Pentarctagon** [comment]
These are the first lawsuits made against a new technology which raise novel legal questions. The legality of the output of any image/text generator trained on copyrighted material is plainly not settled law yet.

And if it is ruled as violating copyright, then "we didn't violate copyright, we just added a button that allowed players to violate copyright" isn't a valid argument.

**Comment by db0** [comment]
Again, that's not how things work. Linux was also novel at the time. 

And yes integrating with an api itself is not illegal either. Anymore than integrating a torrent client would be illegal because someone might upload copyrighted content. To think that an obscure Foss game is going to get in trouble over this when billion dollar players are involved is patently absurd. 

Look I'm not gonna belabor my point. If you just don't want ai art in wesnoth whatsoever It's one thing, but there's no need to hide behind legalese. 

I'm not here to waste anyone's time, I'm only interested in improving the FOSS quality based on mutual aid.

**Comment by Pentarctagon** [comment]
Apple, for example, has routinely sued tiny businesses for using fruit in their logo. There's no such thing as "too small to be sued".

**Comment by soliton-** [comment]
> To think that an obscure Foss game is going to get in trouble over this when billion dollar players are involved is patently absurd.

IMO the main focus is not how likely it is that the wesnoth project gets in trouble but what is generally legally allowed and what we feel is morally sensible.

For example there is plenty legal content that we still do not allow on the official addon server. So going by "well technically it's legal" is not a great argument.

What training data is the generated art based on? I couldn't easily find any info on the website about that. If this is based on FOSS licensed data then it sounds interesting.

**Comment by db0** [comment]
It's stable diffusion models. It's not based on Foss images. Whether that's ethical is currently being argued. I believe it is but it sounds like you'll disagree.

**Comment by doofus-01** [comment]
No doubt a lot of people would find this interesting, but I can't support it.  Over the past year, I've tried to think about this generative AI stuff without being bitter or stubborn, but there is an ethical concern (that one often uses prompts or some form of tuning for "in the style of [a particular artist]" is a tell).  We'll all be using "AI" in various digital tools more in the future, often without realizing it or knowing where training data came from, but wholesale generation is deliberate and is different.

Then there's an issue of motivation.  If Wesnoth were to start using GAI like described in the OP, I'd leave and I doubt any new artist would join or contribute.  I'm not saying that as a threat, it's just how I see it, because why bother?  Maybe it would be fine, and the current situation isn't great, so maybe it would even be an improvement by some measures.  But it is worth considering, even if the legal & ethical issues are resolved.

Maybe it's part of Wesnoth's retro charm that its digital assets are "artisanal & hand-crafted".

**Comment by hermestrismi** [comment]
I personally used to be with generated ai art but I changed my mind because
it clearly limits the ability of human creativity.

Le sam. 20 janv. 2024 à 13:44, doofus-01 ***@***.***> a
écrit :

**Comment by Jonathan-Kelly** [comment]
> Then there's an issue of motivation. If Wesnoth were to start using GAI like described in the OP, I'd leave and I doubt any new artist would join or contribute. I'm not saying that as a threat, it's just how I see it, because why bother?

You could perhaps just use it exclusively for tasks you find tedious. Theoretically, that would boost motivation.

**Comment by CelticMinstrel** [comment]
I think most tedious tasks would be better served by a dumb, specialized script or plugin than by an AI.

**Comment by Jonathan-Kelly** [comment]
It depends on what techniques have evolved by the time the legal issues have been clarified by the courts. If an AI could draw sprite frames for "run" animations in other orientations (north, south, northwest) that would be a big help for a task some might consider tedious, at least in some cases.

There is also a question of free time cost. Doing pixels by hand in GIMP, it may cost 1000 hours to draw every frame of every animation, in every orientation, for every advancement of every unit... of just one faction. Doing the same thing in Blender may cost 100-200 hours. Perhaps an AI could boil that cost down to just (the most creative) 10 hours of work.

**Comment by CelticMinstrel** [comment]
> If an AI could draw sprite frames for "run" animations in other orientations (north, south, northwest) that would be a big help for a task some might consider tedious, at least in some cases.

If it _could_, that would certainly be helpful. Current AI isn't capable of doing something like that though.

It's quite a good example of a task where it would be more effective to write a dedicated non-AI tool to do it, probably using something like Dragonbones.

**Comment by db0** [comment]
My suggestion is not about how to add art to wesnoth. It's about providing easy customization per player. There's nothing wrong with allowing the official art to be non-ai only, but this shouldn't prevent giving the option to each player to customize the looks of their units to fit their own personality. This is not something you can ever achieve with a my tool but ai

**Comment by CelticMinstrel** [comment]
> This is not something you can ever achieve with any tool but ai

This is wholly a matter of opinion, and I definitely don't agree with it. In fact it's almost certainly easier to achieve without AI at the present time.

**Comment by db0** [comment]
Sorry but this is a matter of fact. You can't make a system where each player can customize the look of their units to anything they can think of, on demand, without using ai.

**Comment by doofus-01** [comment]
>My suggestion is not about how to add art to wesnoth. It's about providing easy customization per player.

Right, but I'm not seeing much of a difference.  There is probably a legal distinction that I don't want to armchair lawyer over, but either way we'd be generating and distributing images with stable diffusion or at least enabling it.  Whatever policy we'd come up with to let us have our cake and eat it too wouldn't last long.

**Comment by CelticMinstrel** [comment]
> You can't make a system where each player can customize the look of their units to anything they can think of, on demand, without using ai.

You can't make that with any AI that currently exists anyway, so that's kinda a moot point. And if you drop the "on demand" requirement, the art can be made in the normal way.

**Comment by db0** [comment]
> You can't make that with any AI that currently exists anyway

That's not true at all. Taking an existing unit portrait and changing it to be in the different style is downright trivial.

**Comment by Pentarctagon** [comment]
To summarize, at least how I see it currently:
* The copyright status of the output of these models is currently unclear - there isn't yet any legal precedent directly addressing AI generated art nor are there laws addressing this in the places that are relevant for Wesnoth:
  * United States - this is where [SPI](https://www.spi-inc.org/) is located.
  * European Union - the physical server we use is located in Finland, rented from a German company.

Additionally, for storefronts we'd need to have allow AI generated content:
 * Steam - where the majority of people play Wesnoth. Valve recently [updated their guidelines](https://steamcommunity.com/groups/steamworks/announcements/detail/3862463747997849619), and given this proposal would seem to fall under the "Live-Generated" category, which would open up the risk of Wesnoth getting reported to Valve if illegal content is generated.
 * itch.io - the majority of Wesnoth's revenue currently comes from donations via itch.io.
 * Apple - otherwise we can't create signed builds for macOS.
 
---

Also, just from the purely practical, utilitarian perspective: an AI art generator is not able to take the place of an actual person who knows what they're doing, and even if it could eventually be convinced to create acceptable output I certainly wouldn't want to spend hours being a "prompt engineer" to get there. So choosing to add this in and losing the only artist currently contributing to Wesnoth, likely along with multiple other developers (either because they also strongly oppose the use of AI generated content or because they're upset it was forced in anyway), would likely result in development of Wesnoth ceasing in any meaningful way.

**Comment by db0** [comment]
The copyright status is very clear, as it explicitly has to do with duplicating existing art. Copying styles of other people has always been legal. What is not clear is whether the training of ML models should have been allowed or compensated. Not about the output. If anything, courts have already ruled that the output is in the public domain (i.e. no copyrights).

About the storefronts: 

* Steam: Again, I don't suggest adding AI art directly into Wesnoth, just using a way to generate/modify art during runtime. If necessary this could be disabled on the steam distribution.  Same for apple.
* Itch.io allows it. My own clients and games are on it.

About just not wanting to support Generative AI whatsoever: That is your prerogative of course. If that's the main reason, there's no need to beat around the bush with legalese.

**Comment by Pentarctagon** [comment]
If the content a model is trained on is ruled to be violating copyright, it seems unlikely that the output of such a model will still nevertheless be considered as being perfectly fine.

**Comment by db0** [comment]
No, that's not how copyright works. Copyright is only about the specific expression with some (disney-lobbied) extensions for unique characters.

**Comment by CelticMinstrel** [comment]
That seems like a stretch, given that there have been recorded cases of models spitting out almost direct copies of training images up to noise. A model trained on copyrighted images that can't guarantee it won't do that can't really be treated as copyright-safe.

**Comment by db0** [comment]
For a model to spit out copyrighted characters, it has to be prompted to do so. The fact that it has the *potential* to do so doesn't make it a problem. This has been settled since the betamax trials. If this was a problem, then things like photoshop would also be illegal.
